package net.firefoxsalesman.dungeonsmobs.entity.projectiles;

import net.firefoxsalesman.dungeonsmobs.client.particle.ModParticleTypes;
import net.firefoxsalesman.dungeonsmobs.entity.ModEntities;
import net.minecraft.core.particles.ParticleOptions;
import net.minecraft.world.damagesource.DamageSource;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.projectile.AbstractHurtingProjectile;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.EntityHitResult;
import net.minecraft.world.phys.Vec3;

public class WindcallerBlastProjectileEntity extends AbstractHurtingProjectile {

	public int lifeTime;

	public WindcallerBlastProjectileEntity(Level world) {
		super(ModEntities.WINDCALLER_BLAST_PROJECTILE.get(), world);
		this.setNoGravity(true);
	}

	public WindcallerBlastProjectileEntity(EntityType<? extends WindcallerBlastProjectileEntity> type,
			Level world) {
		super(type, world);
		this.setNoGravity(true);
	}

	public WindcallerBlastProjectileEntity(Level world, LivingEntity shooter, double offsetX, double offsetY,
			double offsetZ) {
		super(ModEntities.WINDCALLER_BLAST_PROJECTILE.get(), shooter,
				new Vec3(offsetX, offsetY, offsetZ), world);
		this.setNoGravity(true);
	}

	public WindcallerBlastProjectileEntity(Level world, double x, double y, double z, double offsetX,
			double offsetY, double offsetZ) {
		super(ModEntities.WINDCALLER_BLAST_PROJECTILE.get(), x, y, z,
				new Vec3(offsetX, offsetY, offsetZ), world);
		this.setNoGravity(true);
	}

	@Override
	public void baseTick() {
		super.baseTick();

		this.lifeTime++;

		if (!level().isClientSide && this.lifeTime > 10) {
			this.remove(RemovalReason.DISCARDED);
		}
	}

	@Override
	public void tick() {
		super.tick();
		for (int i = 0; i < 3; i++) {
			level().addParticle(this.getTrailParticle(), this.getRandomX(1), this.getRandomY(),
					this.getRandomZ(1), 0.0D, 0.0D, 0.0D);
		}
	}

	@Override
	protected float getInertia() {
		return 1.1F;
	}

	@Override
	protected boolean canHitEntity(Entity p_230298_1_) {
		if (!p_230298_1_.isSpectator() && p_230298_1_.isAlive() && p_230298_1_.isPickable()) {
			Entity entity = this.getOwner();
			if (entity != null && entity == p_230298_1_) {
				return false;
			} else {
				return entity == null || !entity.isPassengerOfSameVehicle(p_230298_1_);
			}
		} else {
			return false;
		}
	}

	@Override
	protected ParticleOptions getTrailParticle() {
		return ModParticleTypes.WIND.get();
	}

	@Override
	protected void onHitEntity(EntityHitResult p_213868_1_) {
		super.onHitEntity(p_213868_1_);
		Entity entity = p_213868_1_.getEntity();

		if (!this.level().isClientSide) {
			entity.getRootVehicle().ejectPassengers();
			if (entity instanceof LivingEntity) {
				double d1 = this.getX() - entity.getX();

				double d0;
				for (d0 = this.getZ() - entity.getZ(); d1 * d1
						+ d0 * d0 < 1.0E-4D; d0 = (Math.random() - Math.random()) * 0.01D) {
					d1 = (Math.random() - Math.random()) * 0.01D;
				}

				((LivingEntity) entity).knockback(2F, d1, d0);
			} else {
				entity.setDeltaMovement(entity.getDeltaMovement()
						.add(this.getDeltaMovement().scale(1.5D).add(0, 0.5, 0)));
			}
			entity.hurtMarked = true;
		}
	}

	public boolean isPickable() {
		return false;
	}

	public boolean hurt(DamageSource p_70097_1_, float p_70097_2_) {
		return false;
	}

	protected boolean shouldBurn() {
		return false;
	}

	public boolean isOnFire() {
		return false;
	}

}
